### ISSUE TITLE
Investigate spaCy/Stanza NLP sidecar for lemmatization, POS tagging, and text analysis

### ISSUE THREAD

**Issue by 3stacks** [heuwels/lector#28]
## Summary

Investigate adding NLP capabilities (lemmatization, POS tagging, sentence segmentation, text difficulty scoring) via spaCy or similar NLP tooling. These would run as a Docker sidecar alongside the existing Ollama container.

## Why

Several planned features depend on linguistic analysis beyond what the LLM provides:
- **Lemmatization** — link inflected forms to root words ("geskaam" → "skaam", "huise" → "huis"). Critical for accurate vocab tracking (#26)
- **Text difficulty scoring** — calculate % known words before reading a text (#26)
- **POS tagging** — better part-of-speech labels than LLM guessing
- **Sentence segmentation** — proper sentence splitting for imported content
- **Proper noun detection** — auto-ignore names and places in vocab tracking
- **Word frequency analysis** — vocabulary profiling for the journal feature (#27)
- **Grammar error categorization** — structured error detection for journal corrections (#27)

### spaCy Afrikaans Support

spaCy has an Afrikaans language class (`af`) but **no trained models**. `spacy.blank("af")` gives tokenization only. Everything else (lemma, POS, NER, parsing) would need to be trained from scratch.

The **UD_Afrikaans-AfriBooms** treebank (~49K tokens) is the only Afrikaans training resource. It includes POS, lemma, and dependency annotations but no NER annotations.

spaCy's trainable `EditTreeLemmatizer` (v3.3+) could be trained on AfriBooms for Afrikaans lemmatization.

### Stanza (Alternative — Better Out of the Box)

Stanford's [Stanza](https://stanfordnlp.github.io/stanza/) has **pre-trained Afrikaans models** using AfriBooms for:
- Tokenization
- POS tagging
- Lemmatization (neural seq2seq + dictionary ensemble)
- Dependency parsing

This works immediately without any training. Downside: heavier memory footprint than spaCy.

### Existing Benchmarks

Spark NLP's `pos_afribooms` achieves ~73% POS accuracy on the 14 Universal POS tags. The treebank is small, so accuracy ceilings are real. PRON hits 87% F1, but rare tags like X drop to 18% F1.

### Deployment

Runs as a Docker sidecar with FastAPI. Proven pattern:
- [spacy-api-docker-v3](https://hub.docker.com/r/bbieniek/spacyapi/) — pre-built REST API
- Or build a custom image with a trained Afrikaans pipeline

```yaml
services:
  nlp:
    build: ./nlp  # Custom Dockerfile with af model
    restart: unless-stopped
```

## Recommended Approach

A hybrid strategy, using the right tool for each job:

| Task | Tool | Reason |
|---|---|---|
| Tokenization | spaCy `blank("af")` | Fast, works immediately |
| Sentence segmentation | spaCy `Sentencizer` | Rule-based, good enough for standard text |
| POS tagging | Stanza (pre-trained) or spaCy (trained on AfriBooms) | Stanza works out of the box |
| Lemmatization | Stanza or LLM fallback | Stanza has neural lemmatizer; LLM may handle irregular forms better |
| NER / proper nouns | PROPN tag from POS tagger | No Afrikaans NER data exists; POS proxy is sufficient |
| Text difficulty | App-level calculation | % known words from vocab DB — no NLP model needed |
| Grammar error detection | LLM | Too nuanced for rule-based; LLM excels here |

### Why not spaCy alone?

spaCy's strength is its production-grade inference speed and pipeline architecture, but it needs training data. The AfriBooms treebank is small (~49K tokens), so any model trained on it will have accuracy limitations. For a language like Afrikaans with limited NLP resources, a hybrid approach (lightweight NLP for mechanical tasks + LLM for linguistic judgment) is more practical than trying to build a full spaCy pipeline.

### Why not LLM alone?

The LLM is overkill (and slow) for tasks like tokenization, sentence splitting, and frequency counting. These are better handled by fast, deterministic NLP tools.

## Key Resource

The [UD_Afrikaans-AfriBooms treebank](https://universaldependencies.org/treebanks/af_afribooms/index.html) is the foundation for any Afrikaans NLP work. It's the training data behind every existing Afrikaans model (Stanza, Spark NLP, and any future spaCy pipeline).

## Implementation Phases

1. **Evaluate Stanza vs training spaCy** — benchmark both on AfriBooms for lemmatization and POS accuracy
2. **Deploy as Docker sidecar** — FastAPI wrapper, add to docker-compose
3. **Integrate tokenization + sentence segmentation** — use for import pipeline
4. **Add lemmatization** — link inflected forms in vocab tracking
5. **Add text difficulty scoring** — calculate known-word % per lesson using tokenizer + vocab DB
6. **POS tagging** — improve word state display and auto-ignore proper nouns

## References

- [spaCy API](https://spacy.io/api)
- [spaCy Afrikaans lang class](https://github.com/explosion/spaCy/blob/master/spacy/lang/af/__init__.py)
- [spaCy EditTreeLemmatizer](https://spacy.io/api/edittreelemmatizer)
- [Stanza Afrikaans models](https://stanfordnlp.github.io/stanza/available_models.html)
- [UD Afrikaans-AfriBooms](https://universaldependencies.org/treebanks/af_afribooms/index.html)
- [AfriBooms paper](https://aclanthology.org/W17-0405.pdf)
- [Spark NLP pos_afribooms](https://sparknlp.org/2021/04/06/pos_afribooms_af.html)
- [spacy-api-docker-v3](https://spacy.io/universe/project/spacy-api-docker-v3)
